### ISSUE TITLE
Cannot test XCode 11 project with Swift package dependent on swift nio

### ISSUE THREAD

**Issue by adam-fowler** [apple/swift-nio#1128]
### Expected behavior
If I add a swift package dependent on swift-nio in my project using new XCode 11 swift package functionality I expect to be able to compile for testing. 

### Actual behavior
When I build for testing I get the following error. Missing required module ‘CNIOAtomics’

### Steps to reproduce

1. Create new XCode project
2. Go to File -> Swift Package -> Add Package Dependency
3. Select AsyncHTTPClient
4. Add “import AsyncHTTPClient” at top of AppDelegate.swift
5. Build for testing

If I build for running it works fine. 

### Version
XCode 11 beta 7

**Comment by weissi** [comment]
That is a very reasonable assumption, I don’t think this is a NIO bug though.

@aciidb0mb3r/@neonacho anything special about building an app target for testing with Xcode Package support?

**Comment by weissi** [comment]
@artemredkin you seen this before?

**Comment by aciidb0mb3r** [comment]
This is a known issue that we're tracking internally with <rdar://problem/54587458>. You can work around it by explicitly passing the modulemap flag: 

Add -Xcc -fmodule-map-file=$(PROJECT_TEMP_ROOT)/GeneratedModuleMaps/macosx/<missing module name>.modulemap to OTHER_SWIFT_FLAGS in the test target.

**Comment by adam-fowler** [comment]
Cheers guys, I didn’t really think it was an NIO bug, but thought you guys would like to know as this impacts a lot of projects using your codebase. 

@aciidb0mb3r Any idea on a timeline for a fix for the issue?

**Comment by weissi** [comment]
@adam-fowler yes, thanks very much for letting us know.

**Comment by weissi** [comment]
I'll still close this for now and started tracking the radar Ankit mentioned

**Comment by betzerra** [comment]
Subscribing to this so I know when the radar is fixed. The workaround worked like a charm BTW :-)

**Comment by aciidb0mb3r** [comment]
This should be fixed in Xcode 11.2 Beta 2. Please let me know if it's still reproducing.

**Comment by betzerra** [comment]
@aciidb0mb3r any idea why Xcode 11.2 works and running `xcodebuild -scheme MyAppTests -project ./MyApp.xcodeproj -destination 'platform=iOS Simulator,name=iPhone 8' build test` not?

**Comment by minuscorp** [comment]
Still happening in Xcode 11.2

**Comment by weissi** [comment]
ping @aciidb0mb3r ^^^

**Comment by minuscorp** [comment]
Still happening in Xcode 11.2.1 GM ~~~

**Comment by clementprem** [comment]
Still happening in 11.2

**Comment by AnandWalvekar** [comment]
@aciidb0mb3r 
I am currently using Xcode 11.3.1 & tried adding 
$(PROJECT_TEMP_ROOT)/GeneratedModuleMaps/iphonesimulator/RxCocoaRuntime.modulemap to OTHER_SWIFT_FLAGS in the test target.
Still it keeps giving error

**Comment by weissi** [comment]
Thanks @AnandWalvekar , I updated the Xcode bug report with you info.

**Comment by aciidb0mb3r** [comment]
@AnandWalvekar can you share a sample project that reproduces this issue?

**Comment by jaredsinclair** [comment]
I've run into this on several similar projects, with Swift Packages that contain clang submodules to expose code to the parent Swift module. I've found a workaround that fixes this, at least as of Xcode 11.3.1:

1. In your unit or UI test target, add the parent module (the Swift Package library) to the linked libraries build phase, even if that test target doesn't use it directly.

2. *Sometimes Required* If you are also using an Xcode scheme that's pointed at a unit or UI test target, make sure that the "Build" section of the scheme editor includes the application target in that list. By default, if you add a new Xcode scheme for, e.g., a UI test target, the app won't be included there and so it isn't built properly.

In all reproduced cases, the underlying error is a missing `-fmodule-map-file` flag for any file in the app target that has an `import MyLibrary` statement.

**Comment by jaredsinclair** [comment]
[Missing-Submodule-SPM-Error.zip](https://github.com/apple/swift-nio/files/4227554/Missing-Submodule-SPM-Error.zip)

**Comment by jaredsinclair** [comment]
^^^ sample project reproducing the issue.

To reproduce, choose the UI test **scheme** and CMD+U to run the tests.

**Comment by aciidb0mb3r** [comment]
I can reproduce the issue, thanks a lot for the sample project!

**Comment by trungutt** [comment]
Works fine for the Xcode 11.4, no need for workaround anymore

**Comment by trungutt** [comment]
It seems to work without the workaround on XCode 11.4 @aciidb0mb3r

**Comment by Onno113** [comment]
With XCode 12.0 I still see this compiler error for UI test targets as already perfectly described in an earlier comment: https://github.com/apple/swift-nio/issues/1128#issuecomment-588481616

I can also still reproduce the compiler error it with Xcode 12.0 with the example project running CMD + U with the UI test scheme from the comment: https://github.com/apple/swift-nio/issues/1128#issuecomment-588483372 

Basically follow the comments from @jaredsinclair and you can still reproduce it.

**Comment by MaxDesiatov** [comment]
Still happening with Xcode 12.2 and Xcode 12.3 for me. Here's a test run where it appears https://github.com/TokamakUI/TokamakVapor/runs/1637655163

**Comment by weissi** [comment]
@MaxDesiatov I just opened your test run and from what I can tell, the failure is

```
[1639/1639] Linking TokamakVaporPackageTests
1633
objc[7127]: CLASS: class 'TokamakVaporTests.TokamakVaporTests' 0x10ab02870 small method list 0x10a932c30 is not in immutable memory
1634
Exited with signal code 6
1635
Error: Process completed with exit code 1.
```

which looks like a _different_ compiler bug, no?

FWIW, this bug (`error: Missing required module ‘CNIOAtomics’`) is supposed to be fixed in Xcode 11.3.1.

@MaxDesiatov Can you confirm that the bug you're facing is the small method list error followed by a compiler crash?

**Comment by MaxDesiatov** [comment]
Yes, that's correct. Apologies for spamming this issue then. I think I was redirected here from somewhere else that described having "small method list" error and linked to here as something that helped in resolving that error.

**Comment by weissi** [comment]
@MaxDesiatov don't apologise! Do you happen to already have a Swift bug (Feedback reporter or bugs.swift.org) or Xcode bug (Feedback reporter) that I could look into?

**Comment by MaxDesiatov** [comment]
Thanks! I've filed https://bugs.swift.org/browse/SR-14013.

**Comment by pwerry** [comment]
We've hit this issue again when `swift-nio` pulled in `swift-atomics`. In our case we hit this when we have an embedded app inside a host app, where both depend on the same generated `grpc` files (and thus the same version of `grpc-swift`). 

Except this time the module it's complaining about isn't actually generated at all:
`Missing required module '_AtomicsShims'`

Any hints?

**Comment by Lukasa** [comment]
Where is the error coming from?

**Comment by pwerry** [comment]
The error is an Xcode build error.

We think this might have been an issue with Xcode's build planner. When it builds embedded apps it runs into the same sorts of issues that Ankit mentioned with XCTest targets. Presumably this is an ordering thing. We had solved all the previous missing module errors (CNIOAtomics for example) with `-fmodule-map-file` directives, but for some reason there is no module map generated for _AtomicsShims. 

But we actually discovered an even better way to solve all of these issues - just include the embedded app as a "Target Dependency" in Xcode's Target Dependency build phase. 

We still have no idea why this works but it feels like a possible Xcode bug?

At any rate Luka this doesn't look like an issue to solve with grpc-swift or swift-nio. Hopefully others find this useful!

**Comment by neonichu** [comment]
It looks like the original issue was resolved in 11.3.1. 

@pwerry I'm assuming you are seeing this with a more recent Xcode, the issue is probably at least slightly different and we'd appreciate a new bug report. Since this is an Xcode issue, it would be better to be done through https://feedbackassistant.apple.com

**Comment by pwerry** [comment]
Hi Boris!

FB11741629

Example project here for those interested: https://github.com/pwerry/EmbeddedAppGRPCSwiftExample

Git log will take you back in time to build failures.
